src/java/voldemort/server/VoldemortServer.java

```diff
@@ -516,6 +521,22 @@ public void goOffline() {
 
     public void goOnline() {
         getMetadataStore().setOfflineState(false);
+        List<StorageEngine<ByteArray, byte[], byte[]>> storageEngines =
+                storageService.getStoreRepository().getStorageEnginesByClass(ReadOnlyStorageEngine.class);
+
+        if (storageEngines.size() > 0) {
+            logger.info("Will attempt to removeRemoteObsoleteState() for " + storageEngines.size() + " Read-Only stores.");
+            for (StorageEngine storageEngine: storageEngines) {
+                StoreVersionManager storeVersionManager =
+                        (StoreVersionManager) storageEngine.getCapability(StoreCapabilityType.DISABLE_STORE_VERSION);
+                try {
+                    storeVersionManager.removeRemoteObsoleteState();
+                } catch (Exception e) {
+                    logger.error("Failed to removeRemoteObsoleteState() for store: " + storageEngine.getName(), e);
+                }
+            }
+        }
```

Comment:
I think that there is an opportunity for optimization, but I'm not sure that it's worth it, since this is a rare operation.

I don't understand how you arrive at your 30 minutes figure. We only need to bring back one node from offline to online, not 30.